### DISCUSSION TITLE
How the DART SDK version is checked / Which DART version to use ?

### DISCUSSION THREAD

**Discussion by fmteau** [insideapp-oss/sonar-flutter#226] [Q&A]
Hello all,
When I use your plugin to run on some dart/flutter code, I get at some point in the log, the following message:
`WARN: Could not determine Dart version from output: '' - defaulting to legacy mode`
However, in the same shell, I can run: `dart --version` and get an output (see below), ~~which by the way does not match the regex (on line 235) in AnalyzerExecutable.java.~~
Output: `Dart SDK version: 3.4.3 (stable) (Tue Jun 4 19:51:39 2024 +0000) on "linux_x64"`
I am currently running the check on Windows/WSL (this will be a real Linux machine when integrated within our CI/CD tool)
Any idea what is going on?
Best regards

PS: uname retuns "Linux"

**Comment by Usaid-Dev** [comment]
i think so your dart sdk is not properly defined

**Reply by fmteau** [reply]
How so? That is, I am willing to accept something is amiss, but why am I able to run dart --version on the command line, while the plugin cannot? Does it need a special environment variable or a specif path ? 
Oh, and sorry, the RegEx is ok, This is really the call to dart --version that fails

**Reply by Usaid-Dev** [reply]
Here's a step-by-step guide to set up Flutter and Dart on a Linux system check this all that from this you missed something or not:

### 1. **System Requirements**
Ensure your system meets the minimum requirements:
- Operating System: Linux (64-bit)
- Disk Space: 600 MB (does not include disk space for IDE/tools).
- Tools: `bash`, `curl`, `git 2.x`, `mkdir`, `rm`, `unzip`, `which`

### 2. **Install Flutter**

#### a. Download the Flutter SDK
1. Open a terminal.
2. Download the latest stable release of Flutter SDK:
   ```bash
   cd ~/Downloads
   curl -LO https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_2.10.4-stable.tar.xz
   ```

#### b. Extract the Flutter SDK
1. Extract the downloaded file:
   ```bash
   tar xf flutter_linux_2.10.4-stable.tar.xz
   ```
2. Move the extracted directory to your desired location (e.g., `~/development`):
   ```bash
   mkdir -p ~/development
   mv flutter ~/development
   ```

#### c. Update your PATH
1. Add Flutter to the PATH environment variable:
   ```bash
   export PATH="$PATH:~/development/flutter/bin"
   ```
2. To make this change permanent, add the above line to your `~/.bashrc` or `~/.zshrc` file:
   ```bash
   echo 'export PATH="$PATH:~/development/flutter/bin"' >> ~/.bashrc
   source ~/.bashrc
   ```

### 3. **Install Dependencies**
Run the following command to install necessary dependencies:
```bash
sudo apt-get update
sudo apt-get install -y bash curl git unzip xz-utils zip libglu1-mesa
```

### 4. **Verify Flutter Installation**
1. Run the `flutter doctor` command to check for any dependencies you may still need to install:
   ```bash
   flutter doctor
   ```
2. Follow any additional instructions provided by the `flutter doctor` command to complete the setup.

### 5. **Set Up an Editor**
You can use any text editor or IDE of your choice. Two popular options are:

- **Visual Studio Code**
  1. Install Visual Studio Code:
     ```bash
     sudo snap install --classic code
     ```
  2. Install the Flutter and Dart plugins from the Extensions marketplace.

- **Android Studio**
  1. Download and install Android Studio from [the official website](https://developer.android.com/studio).
  2. During the installation, ensure the boxes for "Android SDK," "Android SDK Platform," and "Android Virtual Device" are checked.
  3. After installation, open Android Studio, go to the "Plugins" section, and install the Flutter and Dart plugins.

### 6. **Set Up Android Device (Optional)**
If you want to run and test your Flutter apps on an Android device:

1. Enable Developer Mode on your Android device and enable USB debugging.
2. Connect your device to your computer via USB.
3. Run the following command to verify if the device is connected:
   ```bash
   flutter devices
   ```

### 7. **Create a New Flutter Project**
1. Create a new Flutter project:
   ```bash
   flutter create my_app
   ```
2. Navigate to the project directory:
   ```bash
   cd my_app
   ```

### 8. **Run Your Flutter App**
1. Ensure an emulator is running or a physical device is connected.
2. Run the app:
   ```bash
   flutter run
   ```

This should get Flutter and Dart set up on your Linux system and ready for development.

**Reply by fmteau** [reply]
I did this... See below command output:
flutter doctor -v
[✓] Flutter (Channel stable, 3.22.2, on Ubuntu 22.04.4 LTS 5.15.153.1-microsoft-standard-WSL2, locale C.UTF-8)
    • Flutter version 3.22.2 on channel stable at /home/iron/development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 761747bfc5 (5 weeks ago), 2024-06-05 22:15:13 +0200
    • Engine revision edd8546116
    • **Dart version 3.4.3**
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at /home/iron/android
    • Platform android-35, build-tools 35.0.0
    • ANDROID_HOME = /home/iron/android
    • ANDROID_SDK_ROOT = /home/iron/android
    • Java binary at: /snap/android-studio/157/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✗] Chrome - develop for the web (Cannot find Chrome executable at google-chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✗] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 14.0.0-1ubuntu1.1
    • cmake version 3.22.1
    ✗ ninja is required for Linux development.
      It is likely available from your distribution (e.g.: apt install ninja-build), or can be downloaded from
      https://github.com/ninja-build/ninja/releases
    • pkg-config version 0.29.2
    ✗ GTK 3.0 development libraries are required for Linux development.
      They are likely available from your distribution (e.g.: apt install libgtk-3-dev)

[✓] Android Studio (version 2023.3)
    • Android Studio at /snap/android-studio/157
    • Flutter plugin version 80.0.1
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] Connected device (1 available)
    • Linux (desktop) • linux • linux-x64 • Ubuntu 22.04.4 LTS 5.15.153.1-microsoft-standard-WSL2

[✓] Network resources
    • All expected network resources are available.

And I used android-studio to install the flutter and dart plugin:
Flutter: 80.0.1 Google
Dart: 233.15271 JetBrains

**Reply by Usaid-Dev** [reply]
Note the following issues that need to be resolved before starting the coding process. Ensure that all items are checked and green, otherwise, you will encounter errors:

- [✗] Chrome - Develop for the web (Cannot find Chrome executable at google-chrome)
  - ! Cannot find Chrome. Try setting `CHROME_EXECUTABLE` to a Chrome executable.

- [✗] Linux toolchain - Develop for Linux desktop
  - Ubuntu clang version 14.0.0-1ubuntu1.1
  - cmake version 3.22.1
  - ✗ Ninja is required for Linux development. It is likely available from your distribution (e.g.: `apt install ninja-build`), or can be downloaded from [ninja-build releases](https://github.com/ninja-build/ninja/releases).
  - pkg-config version 0.29.2
  - ✗ GTK 3.0 development libraries are required for Linux development. They are likely available from your distribution (e.g.: `apt install libgtk-3-dev`).

Always make sure these items are resolved and green before beginning any coding work to avoid errors.

**Reply by fmteau** [reply]
As I want to run only Sonarqube (I am an architect, not a develop), I do think I do not need them (Chrome at least definitively). Now, if you tell me I need the Linux toolchain to run SonarQube, I would be surprised, but I can try to install it to see if this solve the problem.
Anyway, just to be sure I installed the Linux toolchain dependencies, and as expected, SonarQube dart --version fails:
11:26:53.907 DEBUG: Command 'dart --version' finished (exit 0)
11:26:53.907 WARN: Could not determine Dart version from output: '' - defaulting to legacy mode
11:26:53.907 DEBUG: Automatically detected Analyzer output mode: LEGACY
11:26:53.914 INFO: Running 'dart [analyze, --no-fatal-warnings]'
11:26:55.253 INFO: Command 'dart analyze --no-fatal-warnings' finished (exit 0)
11:26:55.254 DEBUG: Cleaned up temporary analysis_options.yaml file

So, obvious my dart installation is ok as the pluging is able to run dart analyze... 
Apparenty, this is the capture of the dart --version output that does not work

**Reply by Usaid-Dev** [reply]
Also try these things:- 

**Update SonarQube and Dart Plugin:**
   Ensure you are using the latest versions of SonarQube and the Dart plugin. Sometimes bugs are fixed in newer releases.

**Configure Environment Variables:**
   Ensure that the environment variables for Dart are correctly set. Add Dart to your PATH if it’s not already there:
   ```bash
   export PATH="$PATH:/path/to/dart-sdk/bin"
   ```

**Check SonarQube Logs:**
   Examine SonarQube logs for any additional information that might indicate why it’s failing to parse the Dart version. Logs are usually found in the `logs` directory of your SonarQube installation.

**Specify Dart Version Manually:**
   If SonarQube is failing to determine the Dart version, you can try to specify it manually in the configuration. Check if there's an option to set the Dart version explicitly in the SonarQube configuration.

**Ensure Compatibility:**
   Verify that the versions of SonarQube, the Dart plugin, and Dart SDK you are using are compatible with each other. Sometimes, version mismatches can cause unexpected behavior.

**Reinstall Dart SDK:**
   If the issue persists, consider reinstalling the Dart SDK to ensure there are no installation issues:
   ```bash
   sudo apt-get remove dart
   sudo apt-get update
   sudo apt-get install dart
   ```

**Run SonarQube with Debug Mode:**
   Run SonarQube in debug mode to get more detailed logs:
   ```bash
   sonar-scanner -X
   ```
   This might provide more insight into what’s causing the issue.

**Reply by fmteau** [reply]
The above lines "11:26:53.907 DEBUG: Co..." come from running sonar-scanner -X..
And again:
1) I can run dart --version without any issue (so dart is in the patth)
2) Sonar can run dart analyze without any issue.

So from these two lines:
11:26:53.907 DEBUG: Command 'dart --version' finished (exit 0)
11:26:53.907 WARN: Could not determine Dart version from output: '' - defaulting to legacy mode
We can see the command execution is successful (exit 0) but the capture of the command output does not work and thus the plugin gets an empty output.

Here is the plugin code:

```
        if (exitCode != 0) {
            LOGGER.warn("Could not determine Dart version from output: '{}' - defaulting to legacy mode", output);
            return AnalyzerOutput.Mode.LEGACY;
        }

        final Matcher matcher = Pattern.compile("^Dart SDK version: (.*) \\((.*)\\) \\((.*)\\) on (.*)$").matcher(output);
        if (!matcher.matches()) {
            LOGGER.warn("Could not determine Dart version from output: '{}' - defaulting to legacy mode", output);
            return AnalyzerOutput.Mode.LEGACY;
        }
```
